services/src/main/java/org/keycloak/storage/ClientStorageManager.java

```diff
@@ -151,18 +151,32 @@ public ClientModel getClientByClientId(RealmModel realm, String clientId) {
 
     @Override
     public Stream<ClientModel> searchClientsByClientIdStream(RealmModel realm, String clientId, Integer firstResult, Integer maxResults) {
-        return query(p -> p.searchClientsByClientIdStream(realm, clientId, firstResult, maxResults), realm, firstResult, maxResults);
+        return query((p, f, m) -> p.searchClientsByClientIdStream(realm, clientId, f, m), realm, firstResult, maxResults);
     }
 
     @Override
     public Stream<ClientModel> searchClientsByAttributes(RealmModel realm, Map<String, String> attributes, Integer firstResult, Integer maxResults) {
-        return query(p -> p.searchClientsByAttributes(realm, attributes, firstResult, maxResults), realm, firstResult, maxResults);
+        return query((p, f, m) -> p.searchClientsByAttributes(realm, attributes, f, m), realm, firstResult, maxResults);
+    }
+
+    @FunctionalInterface
+    interface PaginatedQuery {
+        Stream<ClientModel> query(ClientLookupProvider provider, Integer firstResult, Integer maxResults);
     }
 
-    protected Stream<ClientModel> query(Function<ClientLookupProvider, Stream<ClientModel>> performQuery, RealmModel realm, Integer firstResult, Integer maxResults) {
+    protected Stream<ClientModel> query(PaginatedQuery paginatedQuery, RealmModel realm, Integer firstResult, Integer maxResults) {
         if (maxResults != null && maxResults == 0) return Stream.empty();
 
-        Stream<ClientLookupProvider> providersStream = Stream.concat(Stream.of(session.clientLocalStorage()), getEnabledStorageProviders(session, realm, ClientLookupProvider.class));
+        Set<ClientLookupProvider> extStorageProviders = getEnabledStorageProviders(session, realm, ClientLookupProvider.class).collect(Collectors.toSet());
+        boolean hasExtStorageProviders = !extStorageProviders.isEmpty();
+
+        // when there are external providers involved, we can't do pagination at the lower data layer as we don't know
+        // how many results there will be; i.e. we need to fetch all the clients that match the criteria specified by the query
```

Comment:
This is not true. We are not fetching all clients. Streams should take care of fetching only as many as needed. This means that if we find enough clients in localStorage, there should be no call to external providers.

Reply:
But we don't know how many clients were fetched from the local storage. Not without consuming the stream. If I understand it correctly, this was the whole point of the streamification – to have the results only "on demand" in the stream, we don't even need to process the  whole stream.

Reply:
Well that's the point, we don't need to know how many clients were fetched from local storage.

Reply:
Yes, and therefore we don't know if we need the external storage, if there were enough results in the local storage.

I take it we try to do a best effort solution. We there's only local storage involved, we can do a (potentially) more optimized pagination at the lower data layer. If there are external storages involved, we need to do the pagination at this layer (similarly to users).

Reply:
But maybe we could keep it simple and rely only on the pagination at this layer. I just wanted to use the pagination at the lower layer when possible. ;)

Reply:
You are right. This way it is optimized for cases, when there is no other providers than `localStorage()`. This can be optimized even further you can just call `return paginatedQuery(localStorage, firstResult, maxResults)` when there is not any external provider. 

Also I am not sure whether collecting all storage providers from storage is the best way to check if there is any, maybe we can optimize this as well, but this is just a nit as there probably won't thousands of providers. 

On the other hand I am not sure we will be able to use this optimization in the future as Tree storage won't have local storage anymore.

Reply:
> Also I am not sure whether collecting all storage providers from storage is the best way to check if there is any, maybe we can optimize this as well, but this is just a nit as there probably won't thousands of providers.

Apparently, we probably need to [iterate over all providers](https://github.com/keycloak/keycloak/blob/375a9219d6573a244cc05446616e4d2d715ee097/services/src/main/java/org/keycloak/storage/ClientStorageManager.java#L116-L118) in any case, so I think it's better to fetch them only once.

> On the other hand I am not sure we will be able to use this optimization in the future as Tree storage won't have local storage anymore.

I think with all the current pressure around clients performance it's better to use any optimization possible. But I'll leave it up to you and @hmlnarik on what you agree. ;)

Reply:
> > Also I am not sure whether collecting all storage providers from storage is the best way to check if there is any, maybe we can optimize this as well, but this is just a nit as there probably won't thousands of providers.
> 
> Apparently, we probably need to [iterate over all providers](https://github.com/keycloak/keycloak/blob/375a9219d6573a244cc05446616e4d2d715ee097/services/src/main/java/org/keycloak/storage/ClientStorageManager.java#L116-L118) in any case, so I think it's better to fetch them only once.

Currently yes, because you are collecting the whole stream. When you not collect the stream, the code you sent is done only on those that are queried for clients at the end. If you find enough clients in local storage, non of the code you sent would be executed.

> 
> > On the other hand I am not sure we will be able to use this optimization in the future as Tree storage won't have local storage anymore.
> 
> I think with all the current pressure around clients performance it's better to use any optimization possible. But I'll leave it up to you and @hmlnarik what you agree on. ;)

The issue with storage providers is probably not so important. I think we can leave it as is. However, I would add something like this:
```java
if (!hasExtStorageProviders) {
    return paginatedQuery.query(session.clientLocalStorage(), firstResult, maxResult);
}
```
and remove usage of `firstResultLocal` as it makes the code more intuitive. WDYT?

Reply:
@vmuzikar Maybe you already now, but just to be on the same page with streams, here is some example of what code is called and when:
```
public static void main(String[] args) {
        Stream.of(1,2,3,4,5,6,7,8)
                .peek(i -> System.out.println("Provider number " + i + " was visited."))
                .filter(i -> {
                    System.out.println("Filtering provider " + i);
                    return i % 2 == 0;
                })

                .map(i -> (Supplier<Stream<String>>) () -> {
                    System.out.println("Creating provider number " + i);
                    return Stream.of(1,2,3)
                            .map(j -> "Provider-" + i + " client " + j);
                })

                .flatMap(Supplier::get)

                .skip(0)
                .limit(5)

                .forEach(System.out::println);
    }
```

Now you can change values of skip and limit, and see what methods are called. You can see that in this case, code in peek and filter is not called for providers `>=5` because enough _clients_ were found in previous providers. Also clients are not queried in case provider has uneven number.

Reply:
I did the suggested changes and fixed (hopefully last) failing tests.